# You can copy this entire cell

In [6]:
from sklearn import preprocessing
import pandas as pd
import numpy as np

# Loading Dataset
song_data = pd.read_csv("dataset.csv", sep = ',')

# Trimming Dataset
song_data = song_data[song_data['Genre'] != 'n-a']
song_data = song_data[song_data.iloc[:, 0].str.contains("Global") == True].reset_index(drop=True)

# One-Hot Encoding for Artist and Title
enc = preprocessing.OneHotEncoder()
X_PreEnc = song_data[['Title', 'Artist']]
enc.fit(X_PreEnc)
X_Enc = enc.transform(X_PreEnc).toarray()
X_DF = pd.DataFrame(X_Enc, columns=enc.get_feature_names_out(X_PreEnc.columns))

# Filtering One-Hot Encoding for Artist Only
X_DF2 = X_DF.filter(regex='Artist_').reset_index(drop=True)

# Chosen Features (Can change based on feature selection experiments)
X4 = song_data[['Explicit', 'danceability', 'energy', 'loudness', 'speechiness', 'instrumentalness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics', 'Popu_max']]

# Standard Scaler Preprocessing on chosen features
xcols = ['Explicit', 'danceability', 'energy', 'loudness', 'speechiness', 'instrumentalness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics', 'Popu_max']
normalized_X = X4.values
std_scalor = preprocessing.StandardScaler()
X_scaled = std_scalor.fit_transform(normalized_X)
normalized_X = pd.DataFrame(X_scaled, columns=xcols)

# Combining both DataFrames (One-Hot Encoding Artists + Preprocessed numerical features)
XT = pd.concat([normalized_X.reset_index(drop=True), X_DF2.reset_index(drop=True)], axis=1)

# Target Labels
y = song_data[['Genre_new']].reset_index(drop=True)

# Combining some genres into larger groups (Can add more here if necessary)
# replace_dict = {
#     'rap':'hip hop',
#     'metal':'rock',
#     'trap':'hip hop',
#     'house': 'dance/electronic'
# }

# ybetter = y['Genre_new'].replace(replace_dict)

# YDF = pd.DataFrame(ybetter)
YDF = pd.DataFrame(y)

# All Done with the Dataset
# The X values are in XT
# The y values are in YDF

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (7,8,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [7]:
song_data.isnull().values.any()

True

In [8]:
YDF['Genre_new'].value_counts()

pop                 1706
hip hop             1427
rap                  703
latin                348
rock                 199
r&b/soul             180
dance/electronic     148
house                122
k-pop                119
boy band              86
metal                 81
else                  68
indie                 66
trap                  57
funk                  55
country               28
reggaeton             28
bolero                 2
Name: Genre_new, dtype: int64

## Linear SVM Test

In [413]:
X_train, X_test, y_train, y_test = train_test_split(XT, YDF, test_size=0.2)
clf = LinearSVC(max_iter = 10000, C=1)
#clf = svm.SVC(C=0.1, kernel='linear')
#clf = svm.SVC(C=1, kernel="poly", degree=3)
clf.fit(X_train, y_train.values.ravel())
print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

Training Error: 0.0
Test Error: 0.17603686635944704


Training Error: 0.0

Test Error: 0.18156682027649773

## Logistic Regression Test

In [414]:
clf = LogisticRegression(C=1, solver="lbfgs", penalty="l2", max_iter = 100000000)
clf.fit(X_train,y_train.values.ravel())
#train_accuracy = cross_val_score(clf, X_train2, y_train2).mean()
#test_accuracy = clf.score(X_test2, y_test2)

print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

Training Error: 0.12955278930382663
Test Error: 0.22857142857142854


Training Error: 0.12955278930382663

Test Error: 0.22857142857142854

### 

# Very Messy Experimentation Below

In [13]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn import preprocessing
from sklearn.model_selection import cross_val_score
from sklearn import svm
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors import KNeighborsClassifier

song_data = pd.read_csv("dataset.csv", sep = ',')
song_data = song_data[song_data['Genre'] != 'n-a']
song_data = song_data[song_data.iloc[:, 0].str.contains("Global") == True]
le = preprocessing.LabelEncoder()
X1 = song_data[['Title', 'Artist', 'Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics']]
X = X1
#X = song_data[['Artist', 'Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics']]
X = X.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')
#y = song_data[['Genre']]
y = song_data[['Genre_new']]
y = y.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
#X.head()
X.shape

(5423, 11)

In [155]:
pd.reset_option('max_columns')

In [14]:
X1

Title                     Artist Explicit danceability  \
0             adan y eva               Paulo Londra    False        0.767   
16       it wont kill ya  The Chainsmokers - Louane    False        0.572   
39                  hymn                      Kesha     True        0.488   
58                arrows               Foo Fighters    False        0.515   
80     talking to myself                Linkin Park    False        0.593   
...                  ...                        ...      ...          ...   
83794          lush life               Zara Larsson    False        0.694   
83826       the business                     Tiësto    False        0.798   
83839   scooby doo pa pa                    Dj Kass    False         0.91   
83845          positions              Ariana Grande     True        0.737   
83880  can’t take a joke                      Drake     True        0.878   

      energy loudness liveliness valence    tempo duration_ms acoustics  
0      0.709    -4.47     0.0676    0.72  171.993      258639     0.323  
16      0.53   -8.521      0.127    0.12  170.138      217613    0.0647  
39     0.538   -4.974      0.305    0.38   67.037      205600     0.147  
58     0.917   -7.312      0.145   0.409  121.988      266187  0.000209  
80     0.712   -6.325      0.128   0.473  124.013      231307   0.00184  
...      ...      ...        ...     ...      ...         ...       ...  
83794  0.712   -3.923      0.211   0.799   98.022      201122     0.133  
83826   0.62   -7.079      0.112   0.235  120.031      164000     0.414  
83839  0.754   -2.573      0.122    0.63  120.939      145972    0.0396  
83845  0.802   -4.771     0.0931   0.682  144.015      172325     0.468  
83880  0.432   -9.852     0.0571   0.441   86.956      163973     0.041  

[5423 rows x 11 columns]

In [371]:
enc = preprocessing.OneHotEncoder()

In [213]:
#X2 = X1['Artist'].values.reshape(-1,1)

In [372]:
X2 = X1[['Title','Artist']]

In [373]:
X2

Title                     Artist
0             adan y eva               Paulo Londra
16       it wont kill ya  The Chainsmokers - Louane
39                  hymn                      Kesha
58                arrows               Foo Fighters
80     talking to myself                Linkin Park
...                  ...                        ...
83794          lush life               Zara Larsson
83826       the business                     Tiësto
83839   scooby doo pa pa                    Dj Kass
83845          positions              Ariana Grande
83880  can’t take a joke                      Drake

[5423 rows x 2 columns]

In [374]:
enc.fit(X2)

OneHotEncoder()

In [375]:
enc.categories_

[array(['!', '#proudcatownerremix', '$$$ - with matt ox', ...,
        'è sempre bello', 'échame la culpa', '美女と野獣'], dtype=object),
 array(['*NSYNC',
        '13 Organisé - SCH - Kofs - Jul - Naps - Soso Maness - Elams - Solda - Houari',
        '187 Strassenbande - Bonez MC - Gzuz', ..., 'tha Supreme - Salmo',
        'twocolors',
        'Şanışer - Kamufle - Mert Şenel - Mirac - Aga B - Defkhan - Aspova - Yeis Sensura - Sehabe - Deniz Tekin - Ozbi - Sokrat St - Tahribad-ı İsyan - Beta - Server Uraz - Hayki - Ados - Fuat'],
       dtype=object)]

In [376]:
X2enc = enc.transform(X2).toarray()

In [377]:
X2df = pd.DataFrame(X2enc, columns=enc.get_feature_names_out(X2.columns))

In [378]:
X2df

Title_!  Title_#proudcatownerremix  Title_$$$ - with matt ox  \
0         0.0                        0.0                       0.0   
1         0.0                        0.0                       0.0   
2         0.0                        0.0                       0.0   
3         0.0                        0.0                       0.0   
4         0.0                        0.0                       0.0   
...       ...                        ...                       ...   
5418      0.0                        0.0                       0.0   
5419      0.0                        0.0                       0.0   
5420      0.0                        0.0                       0.0   
5421      0.0                        0.0                       0.0   
5422      0.0                        0.0                       0.0   

      Title_+linda  Title_...ready for it?  Title_0.00  Title_00:00  \
0              0.0                     0.0         0.0          0.0   
1              0.0                     0.0         0.0          0.0   
2              0.0                     0.0         0.0          0.0   
3              0.0                     0.0         0.0          0.0   
4              0.0                     0.0         0.0          0.0   
...            ...                     ...         ...          ...   
5418           0.0                     0.0         0.0          0.0   
5419           0.0                     0.0         0.0          0.0   
5420           0.0                     0.0         0.0          0.0   
5421           0.0                     0.0         0.0          0.0   
5422           0.0                     0.0         0.0          0.0   

      Title_1, 2, 3  Title_1-800-273-8255  Title_1.5  ...  Artist_j-hope  \
0               0.0                   0.0        0.0  ...            0.0   
1               0.0                   0.0        0.0  ...            0.0   
2               0.0                   0.0        0.0  ...            0.0   
3               0.0                   0.0        0.0  ...            0.0   
4               0.0                   0.0        0.0  ...            0.0   
...             ...                   ...        ...  ...            ...   
5418            0.0                   0.0        0.0  ...            0.0   
5419            0.0                   0.0        0.0  ...            0.0   
5420            0.0                   0.0        0.0  ...            0.0   
5421            0.0                   0.0        0.0  ...            0.0   
5422            0.0                   0.0        0.0  ...            0.0   

      Artist_j-hope - Becky G  Artist_jxdn  Artist_salem ilese  \
0                         0.0          0.0                 0.0   
1                         0.0          0.0                 0.0   
2                         0.0          0.0                 0.0   
3                         0.0          0.0                 0.0   
4                         0.0          0.0                 0.0   
...                       ...          ...                 ...   
5418                      0.0          0.0                 0.0   
5419                      0.0          0.0                 0.0   
5420                      0.0          0.0                 0.0   
5421                      0.0          0.0                 0.0   
5422                      0.0          0.0                 0.0   

      Artist_tha Supreme  Artist_tha Supreme - Mara Sattei  \
0                    0.0                               0.0   
1                    0.0                               0.0   
2                    0.0                               0.0   
3                    0.0                               0.0   
4                    0.0                               0.0   
...                  ...                               ...   
5418                 0.0                               0.0   
5419                 0.0                               0.0   
5420                 0.0                        

In [379]:
sub_df = X2df.filter(regex='Artist_')

In [380]:
sub_df

Artist_*NSYNC  \
0               0.0   
1               0.0   
2               0.0   
3               0.0   
4               0.0   
...             ...   
5418            0.0   
5419            0.0   
5420            0.0   
5421            0.0   
5422            0.0   

      Artist_13 Organisé - SCH - Kofs - Jul - Naps - Soso Maness - Elams - Solda - Houari  \
0                                                   0.0                                     
1                                                   0.0                                     
2                                                   0.0                                     
3                                                   0.0                                     
4                                                   0.0                                     
...                                                 ...                                     
5418                                                0.0                                     
5419                                                0.0                                     
5420                                                0.0                                     
5421                                                0.0                                     
5422                                                0.0                                     

      Artist_187 Strassenbande - Bonez MC - Gzuz  \
0                                            0.0   
1                                            0.0   
2                                            0.0   
3                                            0.0   
4                                            0.0   
...                                          ...   
5418                                         0.0   
5419                                         0.0   
5420                                         0.0   
5421                                         0.0   
5422                                         0.0   

      Artist_187 Strassenbande - Bonez MC - RAF Camora - Gzuz - Maxwell  \
0                                                   0.0                   
1                                                   0.0                   
2                                                   0.0                   
3                                                   0.0                   
4                                                   0.0                   
...                                                 ...                   
5418                                                0.0                   
5419                                                0.0                   
5420                                                0.0                   
5421                                                0.0                   
5422                                                0.0                   

      Artist_187 Strassenbande - Gzuz - Bonez MC - LX  \
0                                                 0.0   
1                                                 0.0   
2                                                 0.0   
3                                                 0.0   
4                                                 0.0   
...                                               ...   
5418                                              0.0   
5419                                              0.0   
5420                                              0.0   
5421                                              0.0   
5422                                              0.0   

      Artist_2 Chainz - Ariana Grande  Artist_2 Chainz - Drake - Quavo  \
0                                 0.0                              0.0   
1                                 0.0                              0.0   
2                                 0.0                              0.0   
3                                 0.0                              0.0   
4                              

In [381]:
X2df['Artist_Drake'].sum()

66.0

In [60]:
X1[['Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics']].reset_index()

index Explicit danceability energy loudness liveliness valence    tempo  \
0         0    False        0.767  0.709    -4.47     0.0676    0.72  171.993   
1        16    False        0.572   0.53   -8.521      0.127    0.12  170.138   
2        39     True        0.488  0.538   -4.974      0.305    0.38   67.037   
3        58    False        0.515  0.917   -7.312      0.145   0.409  121.988   
4        80    False        0.593  0.712   -6.325      0.128   0.473  124.013   
...     ...      ...          ...    ...      ...        ...     ...      ...   
5418  83794    False        0.694  0.712   -3.923      0.211   0.799   98.022   
5419  83826    False        0.798   0.62   -7.079      0.112   0.235  120.031   
5420  83839    False         0.91  0.754   -2.573      0.122    0.63  120.939   
5421  83845     True        0.737  0.802   -4.771     0.0931   0.682  144.015   
5422  83880     True        0.878  0.432   -9.852     0.0571   0.441   86.956   

     duration_ms acoustics  
0         258639     0.323  
1         217613    0.0647  
2         205600     0.147  
3         266187  0.000209  
4         231307   0.00184  
...          ...       ...  
5418      201122     0.133  
5419      164000     0.414  
5420      145972    0.0396  
5421      172325     0.468  
5422      163973     0.041  

[5423 rows x 10 columns]

In [129]:
X3 = pd.concat([song_data[['Explicit', 'danceability', 'energy', 'loudness', 'speechiness', 'instrumentalness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics', 'Popu_max']].reset_index(drop=True), X2df.reset_index(drop=True)], axis=1)

In [132]:
pd.reset_option('max_columns')
X3

Explicit danceability energy loudness speechiness instrumentalness  \
0       False        0.767  0.709    -4.47       0.336              0.0   
1       False        0.572   0.53   -8.521      0.0654         0.000169   
2        True        0.488  0.538   -4.974       0.076              0.0   
3       False        0.515  0.917   -7.312      0.0417         0.000476   
4       False        0.593  0.712   -6.325      0.0286              0.0   
...       ...          ...    ...      ...         ...              ...   
5418    False        0.694  0.712   -3.923       0.046              0.0   
5419    False        0.798   0.62   -7.079       0.232           0.0192   
5420    False         0.91  0.754   -2.573       0.239          0.00154   
5421     True        0.737  0.802   -4.771      0.0878              0.0   
5422     True        0.878  0.432   -9.852       0.143              0.0   

     liveliness valence    tempo duration_ms  ... Artist_j-hope  \
0        0.0676    0.72  171.993      258639  ...           0.0   
1         0.127    0.12  170.138      217613  ...           0.0   
2         0.305    0.38   67.037      205600  ...           0.0   
3         0.145   0.409  121.988      266187  ...           0.0   
4         0.128   0.473  124.013      231307  ...           0.0   
...         ...     ...      ...         ...  ...           ...   
5418      0.211   0.799   98.022      201122  ...           0.0   
5419      0.112   0.235  120.031      164000  ...           0.0   
5420      0.122    0.63  120.939      145972  ...           0.0   
5421     0.0931   0.682  144.015      172325  ...           0.0   
5422     0.0571   0.441   86.956      163973  ...           0.0   

      Artist_j-hope - Becky G  Artist_jxdn  Artist_salem ilese  \
0                         0.0          0.0                 0.0   
1                         0.0          0.0                 0.0   
2                         0.0          0.0                 0.0   
3                         0.0          0.0                 0.0   
4                         0.0          0.0                 0.0   
...                       ...          ...                 ...   
5418                      0.0          0.0                 0.0   
5419                      0.0          0.0                 0.0   
5420                      0.0          0.0                 0.0   
5421                      0.0          0.0                 0.0   
5422                      0.0          0.0                 0.0   

      Artist_tha Supreme  Artist_tha Supreme - Mara Sattei  \
0                    0.0                               0.0   
1                    0.0                               0.0   
2                    0.0                               0.0   
3                    0.0                               0.0   
4                    0.0                               0.0   
...                  ...                               ...   
5418                 0.0                               0.0   
5419                 0.0                               0.0   
5420                 0.0                               0.0   
5421                 0.0                               0.0   
5422                 0.0                               0.0   

      Artist_tha Supreme - Marracash  Artist_tha Supreme - Salmo  \
0                                0.0                         0.0   
1                                0.0                         0.0   
2                                0.0                         0.0   
3                                0.0                         0.0   
4                                0.0                         0.0   
...                              ...                         ...   
5418                             0.0                         0.0   
5419                             0.0                         0.0   
5420                             0.0                         0.0   
5421                             0.0                         0.0   
5422         

In [133]:
y = song_data[['Genre_new']].reset_index(drop=True)
y

Genre_new
0              hip hop
1                  pop
2                  pop
3                metal
4                metal
...                ...
5418               pop
5419             house
5420  dance/electronic
5421               pop
5422           hip hop

[5423 rows x 1 columns]

In [134]:
X_train, X_test, y_train, y_test = train_test_split(X3, y, test_size=0.2)

In [135]:
clf = DecisionTreeClassifier(criterion='entropy')
clf.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy')

In [136]:
print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

Training Error: 0.0
Test Error: 0.5576036866359446


In [138]:
clf = LinearSVC(max_iter = 10000, C=1)
#clf = svm.SVC(C=0.1, kernel='linear')
#clf = svm.SVC(C=1, kernel="poly", degree=3)
clf.fit(X_train, y_train.values.ravel())
print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

/Users/jeetajmani/opt/miniconda3/envs/seng/lib/python3.9/site-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Training Error: 0.9866297833102813
Test Error: 0.9861751152073732


In [121]:
X4 = song_data[['Explicit', 'danceability', 'energy', 'loudness', 'speechiness', 'instrumentalness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics', 'Popu_max']].reset_index(drop=True)
X4

Explicit danceability energy loudness speechiness instrumentalness  \
0       False        0.767  0.709    -4.47       0.336              0.0   
1       False        0.572   0.53   -8.521      0.0654         0.000169   
2        True        0.488  0.538   -4.974       0.076              0.0   
3       False        0.515  0.917   -7.312      0.0417         0.000476   
4       False        0.593  0.712   -6.325      0.0286              0.0   
...       ...          ...    ...      ...         ...              ...   
5418    False        0.694  0.712   -3.923       0.046              0.0   
5419    False        0.798   0.62   -7.079       0.232           0.0192   
5420    False         0.91  0.754   -2.573       0.239          0.00154   
5421     True        0.737  0.802   -4.771      0.0878              0.0   
5422     True        0.878  0.432   -9.852       0.143              0.0   

     liveliness valence    tempo duration_ms acoustics  Popu_max  
0        0.0676    0.72  171.993      258639     0.323        10  
1         0.127    0.12  170.138      217613    0.0647        72  
2         0.305    0.38   67.037      205600     0.147       131  
3         0.145   0.409  121.988      266187  0.000209       103  
4         0.128   0.473  124.013      231307   0.00184        86  
...         ...     ...      ...         ...       ...       ...  
5418      0.211   0.799   98.022      201122     0.133       104  
5419      0.112   0.235  120.031      164000     0.414       160  
5420      0.122    0.63  120.939      145972    0.0396        88  
5421     0.0931   0.682  144.015      172325     0.468         1  
5422     0.0571   0.441   86.956      163973     0.041        11  

[5423 rows x 12 columns]

In [118]:
pd.set_option('display.max_columns', None)
song_data.head()

Country                                                Uri  Popularity  \
0   Global  https://open.spotify.com/track/6FyRXC8tJUh863J...    31833.95   
16  Global  https://open.spotify.com/track/5xhJmd0I15jFcEd...      418.00   
39  Global  https://open.spotify.com/track/0wJJhRgdHhT93Z3...       56.00   
58  Global  https://open.spotify.com/track/6JSryEdVJTZq6YB...       78.40   
80  Global  https://open.spotify.com/track/7nAfXgeHfDO50up...      244.95   

                Title                     Artist Album/Single  \
0          adan y eva               Paulo Londra       single   
16    it wont kill ya  The Chainsmokers - Louane        album   
39               hymn                      Kesha       single   
58             arrows               Foo Fighters        album   
80  talking to myself                Linkin Park        album   

                Genre Artist_followers Explicit                   Album  \
0   argentine hip hop       11427104.0    False              Adan y Eva   
16          dance pop       16931568.0    False  Memories...Do Not Open   
39          dance pop        5910376.0     True                    Hymn   
58  alternative metal        8092392.0    False       Concrete and Gold   
80  alternative metal       17132813.0    False          One More Light   

   Release_date Track_number Tracks_in_album danceability energy key loudness  \
0    2018-11-05            1               1        0.767  0.709   1    -4.47   
16   2017-04-07            7              12        0.572   0.53   6   -8.521   
39   2017-08-03            1               1        0.488  0.538   6   -4.974   
58   2017-09-15            7              11        0.515  0.917   1   -7.312   
80   2017-05-19            3              10        0.593  0.712   2   -6.325   

   mode speechiness acoustics instrumentalness liveliness valence    tempo  \
0     1       0.336     0.323              0.0     0.0676    0.72  171.993   
16    0      0.0654    0.0647         0.000169      0.127    0.12  170.138   
39    1       0.076     0.147              0.0      0.305    0.38   67.037   
58    1      0.0417  0.000209         0.000476      0.145   0.409  121.988   
80    0      0.0286   0.00184              0.0      0.128   0.473  124.013   

   duration_ms time_signature Genre_new  Days_since_release  \
0       258639              4   hip hop               740.0   
16      217613              4       pop              1314.0   
39      205600              4       pop              1199.0   
58      266187              4     metal              1156.0   
80      231307              4     metal              1272.0   

    Released_after_2017  Explicit_false  Explicit_true  album  compilation  \
0                   1.0               1              0      0            0   
16                  1.0               1              0      1            0   
39                  1.0               0              1      0            0   
58                  1.0               1              0      1            0   
80                  1.0               1              0      1            0   

    single  bolero  boy band  country  dance/electronic  else  funk  hip hop  \
0        1       0         0        0                 0     0     0        1   
16       0       0         0        0                 0     0     0        0   
39       1       0         0        0                 0     0     0        0   
58       0       0         0        0                 0     0     0        0   
80       0       0         0        0                 0     0     0        0   

    house  indie  jazz  k-pop  latin  metal  opm  pop  r&b/soul  rap  reggae  \
0       0      0     0      0      0      0    0    0         0    0       0   
16      0      0     0      0      0      0    0    1         0    0       0   
39      0      0     0      0      0      0    0    1         0    0       0   
58      0      0     0      0      0      1    0    0         0    0       0   
80      0    

In [120]:
# number 1 songs
song_data[song_data['Popu_max'] == 1]

Country                                                Uri  Popularity  \
1720   Global  https://open.spotify.com/track/3swc6WTsr7rl9Dq...    50735.25   
2953   Global  https://open.spotify.com/track/4HBZA5flZLE435Q...    30763.80   
3628   Global  https://open.spotify.com/track/4R2kfaDFhslZEMJ...    11167.85   
6203   Global  https://open.spotify.com/track/116H0KvKr2Zl4RP...    58972.05   
7654   Global  https://open.spotify.com/track/0dbTQYW3Ad1FTzI...     5866.40   
...       ...                                                ...         ...   
78416  Global  https://open.spotify.com/track/0sf12qNH5qcw8qp...    39297.70   
81557  Global  https://open.spotify.com/track/0b9oOr2ZgvyQu88...    20447.75   
83394  Global  https://open.spotify.com/track/2Fxmhks0bxGSBdJ...   130062.25   
83534  Global  https://open.spotify.com/track/4Sib57MmYGJzSvk...     7320.20   
83845  Global  https://open.spotify.com/track/35mvY5S1H3J2QZy...     3713.40   

                 Title                         Artist Album/Single  \
1720            psycho    Post Malone - Ty Dolla $ign        album   
2953      stuck with u  Ariana Grande - Justin Bieber       single   
3628          cardigan                   Taylor Swift        album   
6203               mia              Bad Bunny - Drake       single   
7654         mona lisa     Lil Wayne - Kendrick Lamar        album   
...                ...                            ...          ...   
78416  blinding lights                     The Weeknd       single   
81557  this is america               Childish Gambino       single   
83394          bad guy                  Billie Eilish        album   
83534              me!    Taylor Swift - Brendon Urie       single   
83845        positions                  Ariana Grande        album   

                           Genre Artist_followers Explicit  \
1720                     dfw rap       29173640.0     True   
2953                   dance pop       52571724.0    False   
3628                   dance pop       33485486.0    False   
6203                       latin       26265604.0    False   
7654                     hip hop        9717614.0     True   
...                          ...              ...      ...   
78416  canadian contemporary r&b         25562055    False   
81557                atl hip hop          7079534     True   
83394                 electropop         34697267    False   
83534                  dance pop         33485486    False   
83845                  dance pop         52571724     True   

                                          Album Release_date Track_number  \
1720                       beerbongs & bentleys   2018-04-27            8   
2953                               Stuck with U   2020-05-08            1   
3628                                   folklore   2020-07-24            2   
6203                          MIA (feat. Drake)   2018-10-11            1   
7654                               Tha Carter V   2018-09-28            8   
...                                         ...          ...          ...   
78416                           Blinding Lights   2019-11-29            1   
81557                           This Is America   2018-05-06            1   
83394  WHEN WE ALL FALL ASLEEP, WHERE DO WE GO?   2019-03-29            2   
83534                                       ME!   2019-04-26            1   
83845                                 Positions   2020-10-30           12   

      Tracks_in_album danceability energy key loudness mode speechiness  \
1720               18         0.75   0.56   8   -8.094    1       0.105   
2953                1        0.597   0.45   8   -6.658    1      0.0418   
3628               16        0.613  0.581   0   -8.588    0      0.0424   
6203                1        0.817  0.539   6   -6.349    0      0.0621   
7654               23        0.856  0.694   2   -4.393    1       0.169   
...               ...          ...    ...  ..      ...  ...         ...   
78416    

In [235]:
xcols = ['Explicit', 'danceability', 'energy', 'loudness', 'speechiness', 'instrumentalness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics', 'Popu_max']
normalized_X = X4.values
std_scalor = preprocessing.StandardScaler()
X_scaled = std_scalor.fit_transform(normalized_X)
normalized_X = pd.DataFrame(X_scaled, columns=xcols)

In [236]:
XT = pd.concat([normalized_X.reset_index(drop=True), X2df.reset_index(drop=True)], axis=1)

In [239]:
XT

Explicit  danceability    energy  loudness  speechiness  \
0    -0.947355      0.577183  0.432321  0.750346     1.635203   
1    -0.947355     -0.748378 -0.628602 -0.802740    -0.582711   
2     1.055571     -1.319389 -0.581186  0.557121    -0.495831   
3    -0.947355     -1.135849  1.665126 -0.339230    -0.776963   
4    -0.947355     -0.605625  0.450102  0.039170    -0.884335   
...        ...           ...       ...       ...          ...   
5418 -0.947355      0.080947  0.450102  0.960057    -0.741719   
5419 -0.947355      0.787913 -0.095177 -0.249901     0.782790   
5420 -0.947355      1.549261  0.699034  1.477625     0.840164   
5421  1.055571      0.373251  0.983527  0.634948    -0.399115   
5422  1.055571      1.331733 -1.209443 -1.313024     0.053320   

      instrumentalness  liveliness   valence     tempo  duration_ms  ...  \
0            -0.161006   -0.809493  1.072883  1.659786     1.022513  ...   
1            -0.158792   -0.383307 -1.613182  1.597588     0.227903  ...   
2            -0.161006    0.893817 -0.449220 -1.859329    -0.004770  ...   
3            -0.154771   -0.254160 -0.319394 -0.016854     1.168706  ...   
4            -0.161006   -0.376132 -0.032880  0.051043     0.493135  ...   
...                ...         ...       ...       ...          ...  ...   
5418         -0.161006    0.219381  1.426549 -0.820420    -0.091502  ...   
5419          0.090489   -0.490930 -1.098353 -0.082471    -0.810497  ...   
5420         -0.140834   -0.419181  0.669974 -0.052026    -1.159672  ...   
5421         -0.161006   -0.626535  0.902766  0.721699    -0.649255  ...   
5422         -0.161006   -0.884829 -0.176137 -1.191457    -0.811020  ...   

      Artist_j-hope  Artist_j-hope - Becky G  Artist_jxdn  Artist_salem ilese  \
0               0.0                      0.0          0.0                 0.0   
1               0.0                      0.0          0.0                 0.0   
2               0.0                      0.0          0.0                 0.0   
3               0.0                      0.0          0.0                 0.0   
4               0.0                      0.0          0.0                 0.0   
...             ...                      ...          ...                 ...   
5418            0.0                      0.0          0.0                 0.0   
5419            0.0                      0.0          0.0                 0.0   
5420            0.0                      0.0          0.0                 0.0   
5421            0.0                      0.0          0.0                 0.0   
5422            0.0                      0.0          0.0                 0.0   

      Artist_tha Supreme  Artist_tha Supreme - Mara Sattei  \
0                    0.0                               0.0   
1                    0.0                               0.0   
2                    0.0                               0.0   
3                    0.0                               0.0   
4                    0.0                               0.0   
...                  ...                               ...   
5418                 0.0                               0.0   
5419                 0.0                               0.0   
5420                 0.0                               0.0   
5421                 0.0                               0.0   
5422                 0.0                               0.0   

      Artist_tha Supreme - Marracash  Artist_tha Supreme - Salmo  \
0                                0.0                         0.0   
1                                0.0                         0.0   
2                                0.0                         0.0   
3                                0.0                         0.0   
4                                0.0                         0.0   
...                              ...                         ...   
5418                             0.0                         0.0   
5419                             0.0              

In [332]:
y['Genre_new'].unique()

array(['hip hop', 'pop', 'metal', 'indie', 'rap', 'funk', 'r&b/soul',
       'k-pop', 'rock', 'dance/electronic', 'latin', 'trap', 'country',
       'house', 'reggaeton', 'boy band', 'else', 'bolero'], dtype=object)

In [333]:
replace_dict = {
    'rap':'hip hop',
    'metal':'rock',
    'trap':'hip hop',
    'house': 'dance/electronic'
}

ybetter = y['Genre_new'].replace(replace_dict)

In [338]:
YDF = pd.DataFrame(ybetter)

In [340]:
YDF['Genre_new'].unique()

array(['hip hop', 'pop', 'rock', 'indie', 'funk', 'r&b/soul', 'k-pop',
       'dance/electronic', 'latin', 'country', 'reggaeton', 'boy band',
       'else', 'bolero'], dtype=object)

In [366]:
X_train, X_test, y_train, y_test = train_test_split(XT, YDF, test_size=0.2)
clf = DecisionTreeClassifier(criterion='entropy')
clf.fit(X_train, y_train)
print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

Training Error: 0.0
Test Error: 0.43870967741935485


In [355]:
YDF.head(35)

Genre_new
0            hip hop
1                pop
2                pop
3               rock
4               rock
5            hip hop
6            hip hop
7                pop
8              indie
9                pop
10           hip hop
11               pop
12           hip hop
13           hip hop
14           hip hop
15              funk
16          r&b/soul
17               pop
18           hip hop
19           hip hop
20             k-pop
21               pop
22              rock
23               pop
24           hip hop
25           hip hop
26  dance/electronic
27               pop
28           hip hop
29               pop
30               pop
31  dance/electronic
32           hip hop
33           hip hop
34           hip hop

In [261]:
y.values[:10]

array([['hip hop'],
       ['pop'],
       ['pop'],
       ['metal'],
       ['metal'],
       ['hip hop'],
       ['hip hop'],
       ['pop'],
       ['indie'],
       ['pop']], dtype=object)

In [344]:
predictions = clf.predict(XT)

In [345]:
wrongans = (np.where(np.concatenate(YDF.values, axis=0) != predictions))[0]

In [347]:
len(wrongans)

455

In [348]:
wrongans

array([   8,   11,   16,   22,   31,   32,   40,   54,   67,   79,   88,
         92,   96,  111,  133,  142,  143,  157,  160,  164,  172,  176,
        183,  214,  240,  241,  256,  258,  272,  275,  289,  292,  295,
        319,  366,  383,  384,  395,  460,  469,  471,  478,  492,  505,
        510,  562,  565,  574,  576,  593,  610,  631,  649,  650,  653,
        706,  714,  724,  752,  754,  756,  757,  765,  776,  795,  827,
        841,  855,  862,  884,  893,  894,  911,  947,  959,  964,  966,
        972,  975,  978,  979,  997, 1012, 1020, 1021, 1028, 1033, 1055,
       1075, 1078, 1098, 1099, 1170, 1172, 1178, 1185, 1187, 1190, 1192,
       1197, 1198, 1199, 1202, 1207, 1208, 1215, 1218, 1227, 1269, 1295,
       1296, 1304, 1307, 1311, 1328, 1342, 1355, 1356, 1357, 1371, 1378,
       1382, 1399, 1403, 1431, 1434, 1448, 1459, 1473, 1491, 1495, 1503,
       1518, 1544, 1608, 1616, 1646, 1652, 1660, 1664, 1668, 1673, 1683,
       1693, 1701, 1716, 1753, 1779, 1780, 1799, 18

In [354]:
predictions[31]

'boy band'

In [370]:
X_train, X_test, y_train, y_test = train_test_split(XT, YDF, test_size=0.2)
clf = LinearSVC(max_iter = 10000, C=1)
#clf = svm.SVC(C=0.1, kernel='linear')
#clf = svm.SVC(C=1, kernel="poly", degree=3)
clf.fit(X_train, y_train.values.ravel())
print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

Training Error: 0.0
Test Error: 0.16866359447004609


## Trying genre to genre_new experiments

In [149]:
G_DF = song_data[['Genre', 'Explicit']].reset_index(drop=True)
GN_DF = song_data['Genre_new'].reset_index(drop=True)

In [151]:
encG = preprocessing.OneHotEncoder()

encG.fit(G_DF)

G_DF_enc = encG.transform(G_DF).toarray()

G_DF2 = pd.DataFrame(G_DF_enc, columns=encG.get_feature_names_out(G_DF.columns))

In [156]:
G_DF2

Genre_a cappella  Genre_acoustic pop  Genre_adult standards  \
0                  0.0                 0.0                    0.0   
1                  0.0                 0.0                    0.0   
2                  0.0                 0.0                    0.0   
3                  0.0                 0.0                    0.0   
4                  0.0                 0.0                    0.0   
...                ...                 ...                    ...   
5418               0.0                 0.0                    0.0   
5419               0.0                 0.0                    0.0   
5420               0.0                 0.0                    0.0   
5421               0.0                 0.0                    0.0   
5422               0.0                 0.0                    0.0   

      Genre_afro dancehall  Genre_afrofuturism  \
0                      0.0                 0.0   
1                      0.0                 0.0   
2                      0.0                 0.0   
3                      0.0                 0.0   
4                      0.0                 0.0   
...                    ...                 ...   
5418                   0.0                 0.0   
5419                   0.0                 0.0   
5420                   0.0                 0.0   
5421                   0.0                 0.0   
5422                   0.0                 0.0   

      Genre_afrofuturismo brasileiro  Genre_afroswing  Genre_alabama rap  \
0                                0.0              0.0                0.0   
1                                0.0              0.0                0.0   
2                                0.0              0.0                0.0   
3                                0.0              0.0                0.0   
4                                0.0              0.0                0.0   
...                              ...              ...                ...   
5418                             0.0              0.0                0.0   
5419                             0.0              0.0                0.0   
5420                             0.0              0.0                0.0   
5421                             0.0              0.0                0.0   
5422                             0.0              0.0                0.0   

      Genre_albanian hip hop  Genre_album rock  ...  Genre_trap  \
0                        0.0               0.0  ...         0.0   
1                        0.0               0.0  ...         0.0   
2                        0.0               0.0  ...         0.0   
3                        0.0               0.0  ...         0.0   
4                        0.0               0.0  ...         0.0   
...                      ...               ...  ...         ...   
5418                     0.0               0.0  ...         0.0   
5419                     0.0               0.0  ...         0.0   
5420                     0.0               0.0  ...         0.0   
5421                     0.0               0.0  ...         0.0   
5422                     0.0               0.0  ...         0.0   

      Genre_trap latino  Genre_tropical house  Genre_turkish hip hop  \
0                   0.0                   0.0                    0.0   
1                   0.0                   0.0                    0.0   
2                   0.0                   0.0                    0.0   
3                   0.0                   0.0                    0.0   
4                   0.0                   0.0                    0.0   
...                 ...                   ...                    ...   
5418                0.0                   0.0                    0.0   
5419                0.0                   0.0                    0.0   
5420                0.0                   0.0                    0.0   
5421                0.0                   0.0                    0.0   
5422                0.0                   0.0                    0

In [157]:
X_train, X_test, y_train, y_test = train_test_split(G_DF2, GN_DF, test_size=0.2)

clf = DecisionTreeClassifier(criterion='entropy')
clf.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy')

In [162]:
print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

Training Error: 0.0
Test Error: 0.03594470046082954


In [ ]:
### Jeet's Tests
song_data1 = song_data[song_data.iloc[:, 0].str.contains("Global") == True]

#song_data1 = song_data1[song_data1['Genre_new'] != 'n-a']
# Binary Classificati
#song_data1 = song_data1[song_data1.iloc[:, 26].str.contains("hip hop|rap") == True]
le = preprocessing.LabelEncoder()
Xj = song_data1[['Title', 'Artist', 'Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics']]
Xj = Xj.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')
yj = song_data1[['Genre_new']]
yj = yj.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')

X_trainj, X_testj, y_trainj, y_testj = train_test_split(Xj, yj, test_size=0.2)

clfj = DecisionTreeClassifier()
clfj.fit(X_trainj, y_trainj)
print("Training Error: " + str(1.0-clfj.score(X_trainj, y_trainj)))
print("Test Error: " + str(1.0-clfj.score(X_testj, y_testj)))

Training Error: 0.0
Test Error: 0.3419354838709677


In [ ]:
test_errors = []
train_errors = []

for i in range(0,100):
  song_data1 = song_data[song_data.iloc[:, 0].str.contains("Global") == True]

  #song_data1 = song_data1[song_data1['Genre_new'] != 'n-a']
  # Binary Classificati
  song_data1 = song_data1[song_data1.iloc[:, 26].str.contains("hip hop|metal") == True]
  le = preprocessing.LabelEncoder()
  Xj = song_data1[['Title', 'Artist', 'Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics']]
  Xj = Xj.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')
  yj = song_data1[['Genre_new']]
  yj = yj.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')

  X_trainj, X_testj, y_trainj, y_testj = train_test_split(Xj, yj, test_size=0.2)

  clfj = DecisionTreeClassifier()
  clfj.fit(X_trainj, y_trainj)

  train_errors.append(1.0-clfj.score(X_trainj, y_trainj))
  test_errors.append(1.0-clfj.score(X_testj, y_testj))

print("Training Error Average: " + str(np.average(train_errors)))
print("Test Error Average: " + str(np.average(test_errors)))

# Training Error Average: 0.0
# Test Error Average: 0.038741721854304634

Training Error Average: 0.0
Test Error Average: 0.038741721854304634


In [ ]:
song_data1 = song_data[song_data.iloc[:, 0].str.contains("Global") == True]

song_data1 = song_data1[song_data1['Genre'] != 'n-a']
le = preprocessing.LabelEncoder()
Xj = song_data1[['Title', 'Artist', 'Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo']]
Xj = Xj.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')
yj = song_data1[['Genre']]
yj = yj.apply(lambda col: le.fit_transform(col.astype(str)), axis=0, result_type='expand')
X_trainj, X_testj, y_trainj, y_testj = train_test_split(Xj, yj, test_size=0.2)

clfj = DecisionTreeClassifier()
clfj.fit(X_trainj, y_trainj)
print("Training Error: " + str(1.0-clfj.score(X_trainj, y_trainj)))
print("Test Error: " + str(1.0-clfj.score(X_testj, y_testj)))

Training Error: 0.0
Test Error: 0.40737327188940087


In [ ]:
#Normalize data in X 

#Minmax scalor preprocessing
#normalized_X=(X-X.min())/(X.max()-X.min())

#Standard scalor preprocessing
xcols = ['Title', 'Artist', 'Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics']
#xcols = ['Artist', 'Explicit', 'danceability', 'energy', 'loudness', 'liveliness', 'valence', 'tempo', 'duration_ms', 'acoustics']
normalized_X = X.values
std_scalor = preprocessing.StandardScaler()
X_scaled = std_scalor.fit_transform(normalized_X)
normalized_X = pd.DataFrame(X_scaled, columns=xcols)

X_train3, X_test3, y_train3, y_test3 = train_test_split(normalized_X, y, test_size=0.2)

In [ ]:
#Reducing dataset size - not used in cells below anymore
X_train2 = X_train.iloc[0:7000,:]
X_test2 = X_test.iloc[0:1200,:]
y_train2 = y_train.iloc[0:7000,:]
y_test2 = y_test.iloc[0:1200,:]

In [ ]:
#Logistic Refression


#clf = LogisticRegression(multi_class='ovr', max_iter=10000)
clf = LogisticRegression(C=1, solver="lbfgs", penalty="l2", max_iter = 100000000)
clf.fit(X_train,y_train.values.ravel())
#train_accuracy = cross_val_score(clf, X_train2, y_train2).mean()
#test_accuracy = clf.score(X_test2, y_test2)

print("Training Error: " + str(1.0-clf.score(X_train, y_train)))
print("Test Error: " + str(1.0-clf.score(X_test, y_test)))

Training Error: 0.5198248040571691
Test Error: 0.5004608294930876


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of f AND g EVALUATIONS EXCEEDS LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [ ]:
#Support Vector Machine

clf = LinearSVC(max_iter = 10000, C=1)
#clf = svm.SVC(C=0.1, kernel='linear')
#clf = svm.SVC(C=1, kernel="poly", degree=3)
clf.fit(X_train3, y_train3.values.ravel())
print("Training Error: " + str(1.0-clf.score(X_train3, y_train3)))
print("Test Error: " + str(1.0-clf.score(X_test3, y_test3)))

Training Error: 0.5154449054863992
Test Error: 0.5235023041474655


In [ ]:
#Decision tree
clf = DecisionTreeClassifier()
clf.fit(X_train3, y_train3.values.ravel())
print("Training Error: " + str(1.0-clf.score(X_train3, y_train3)))
print("Test Error: " + str(1.0-clf.score(X_test3, y_test3)))

Training Error: 0.0
Test Error: 0.3354838709677419


In [ ]:
#Random Forest # not enough ram to run on full dataset- change params
clf = RandomForestClassifier()
clf.fit(X_train3, y_train3.values.ravel())
print("Training Error: " + str(1.0-clf.score(X_train3, y_train3)))
print("Test Error: " + str(1.0-clf.score(X_test3, y_test3)))

Training Error: 0.0
Test Error: 0.3640552995391705


In [ ]:
clf = KNeighborsClassifier()
clf.fit(X_train3, y_train3.values.ravel())
print("Training Error: " + str(1.0-clf.score(X_train3, y_train3)))
print("Test Error: " + str(1.0-clf.score(X_test3, y_test3)))

Training Error: 0.3886583679114799
Test Error: 0.5612903225806452
